---
title: "nanopoRe: analysis of enrichment performance from targeted sequencing studies"
author: "Stephen Rudd"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{nanopoRe: analysis of enrichment performance from targeted sequencing studies}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
library(nanopoRe)
library(knitr)
library(emojifont)
init()
setOutputFormat("file")
```

## Parsing of configuration file

The cas9-mediated target enrichment workflow is configured by a YAML format file that provides paths to the required files, and parameters specific to the workflow. The YAML file should be loaded and evaluated

```{r loadFile}
yamlFile <- system.file("extdata", "cas9_demo.yaml", package = "nanopoRe")
importConfigYAML(yamlFile=yamlFile)
```

Since we are operating within a vignette - or out of the expected CONDA/Snakemake environment a manual update for a couple of fields makes sense

```{r, results='hide'}
unmappedQ <- system.file("extdata", "cas9_FAK76554.unmapped.quals", package = "nanopoRe")
bamFile <- system.file("extdata", "cas9_FAK76554.bam", package = "nanopoRe")
referenceGenome <- system.file("extdata", "cas9_demo_ref.fasta", package = "nanopoRe")
bedTargets <- system.file("extdata", "cas9_demo_target.bed", package = "nanopoRe")

setCachedYAMLValue(field="reference_genome", value=referenceGenome)
setCachedYAMLValue(field="target_regions", value=bedTargets)
addCachedYAMLValue(field="bam_file", value=bamFile)
addCachedYAMLValue(field="unmapped_quals", value=unmappedQ)
```

```{r}
cachedYAMLToYAML(format="Kable")
```

Our ducks should now be lined up for further analysis and exploration ...

Ensure that the enrichment analysis has been run

```{r, results='hide'}
RunEnrichmentAnalysis(force=TRUE, mc.cores=1)
```

This can also be validated with the **`isEnrichmentAnalysisComplete()`** command

```{r}
isEnrichmentAnalysisComplete()
```


```{r}
importEnrichmentResults()
```


```{r}
include_graphics( enrichmentExecutiveSummary() )
```

```{r}
enrichmentMappingByGenomicSegment()
```

```{r}
enrichmentTargetPerformanceTable()
```


```{r}
enrichmentWriteExcelOnTarget()
```


```{r}
targets <- enrichmentGetTargetList()
include_graphics( enrichmentSingleTargetPlot(targets[[1]]))
```

```{r}
include_graphics( enrichmentStrandedTargetPlot(targets[[1]]) )
```


```{r}
include_graphics( enrichmentMultiGeneCoveragePanel() )
```

```{r}
include_graphics( enrichmentCoverageTypeOverChromosomes() )
```


```{r}
include_graphics( enrichmentOffTargetKaryogram() )
```

```{r}
enrichmentOffTargetTable()
```

```{r}
enrichmentWriteExcelOffTarget()
```




The end ...
